keps/sig-cluster-lifecycle/draft-20180316-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,243 @@
+creating custom clusters.
+
+On the other side, it required a continuous effort to keep this
+component up to date with the changes of `kubeadm init`, especially with regards
+to the proliferation of flags and the duplication of code and, as a consequence, with
+the risk of potential inconsistencies between init and phases implementation logic.
+
+This KEP aims to define the way forward for graduating `kubeadm alpha phase`
+commands to beta (and GA) and, at the same time, to enable a significant reduction
+of the effort for maintaining this feature.
+
+### Goals
+
+* To define the approach for graduating to beta (GA) the `kubeadm phase` user
+  interface.
+* To define a new approach for implementing the `kubeadm phase` user interface
+  that significantly reduces the effort for maintaining this feature up to date
+  with the continuous changes of the init workflow.
+* To provide a centralized point for controlling the proliferation of flags in
+  kubeadm.
+* To avoid inconsistencies between init and phases (both at UX and implementation
+  logic level)
+
+### Non-goals
+
+* This proposal doesn't include any changes/improvement to the init workflow
+  internal logic.
+* Nothing in this proposal should prevent practices that exist today.
+
+### Challenges and Open Questions
+
+* How to reduce the effort for maintaining `kubeadm phase` feature up to date
+  with the continuous changes of the init workflow?
+* How to provide an intuitive user interface for both phases and sub-phases
+  (e.g  `kubeadm alpha phase certs` and `kubeadm alpha phase certs apiserver`)?
+* Should we preserve all the second/third level nested phases implemented today
+  (e.g. certs CA, certs api-server, certs sa etc.), or should we allow only second
+  level phases (e.g certs), or should we maintain & simplify second/third level phases?
+* Should we preserve or deprecate the (few) phases implemented today that are not
+  part of the `kubeadm init` workflow? If preserve, in which form? as a new first
+  level commands?
+* While `kubeadm init` executes some phases conditionally (e.g. self-hosting only
+  if the corresponding feature-gate enable), the current implementation of phases
+  executes them unconditionally. Should we preserve this behavior?
+* How to preserve the possibility to automatically generate kubeadm documentation?
+* How to address/enable the possibility to generate machine readable output from
+  kubeadm commands?
+
+## Proposal
+
+### User Stories
+
+#### Invoke a specific task/phase of the kubeadm init workflow
+
+As a kubernetes administrator, I want to run only a specific task/phase of the
+`kubeadm init` workflow.
+
+### Implementation Details
+
+In order to achieve the goals of this proposal the approach suggested is to customize
+the `kubeadm init` command transforming it's usage from:
+
+```bash
+Usage:
+  kubeadm init [flags]
+```
+
+to
+
+```bash
+Usage:
+  kubeadm init [phases] [flags]
```

Comment:
Are you implying there would be like a comma-separated list of phases or just a name of a single phase?
Arguments like this make it too ambiguous, how about `--only=<listOfPhases>` or `--only=<singlePhase>`?

Reply:
Open for discussion....
> Are you implying there would be like a comma-separated list of phases or just a name of a single phase?
With regards to the number of phases we can limit it to 1 or keep it open. [phases] can be changed in whatever we want accordingly (the current form was inspired by [flags], [args] etc.)
> Arguments like this make it too ambiguous, how about --only=<listOfPhases> or --only=<singlePhase>?
Personally I think that using arguments is the best choice for this use case